File: LeetCode/Java/src/solutions/p0027_RemoveElement/P0027.java
package solutions.p0027_RemoveElement;

public class P0027 {
    public int removeElement(int[] nums, int val) {
        /*
         * Solution v1: Brute Force
         *
         * Runtime: 0 ms @ (beats) 100%
         * Memory Usage: 39.4 MB @ (beats) 11.70%
         */
        // int left = 0, right = nums.length - 1;
        // int ans = 0;
        // while (left <= right) {
        //     if (nums[left] == val) {
        //         while (right >= 0 && nums[right] == val) {
        //             right--;
        //         }
        //         if (right < 0 || right <= left) {
        //             break;
        //         }
        //         int temp = nums[left];
        //         nums[left] = nums[right];
        //         nums[right] = temp;
        //         right--;
        //     }
        //     ans++;
        //     left++;
        // }
        // return ans;

        /*
         * (learn from Internet)
         * Solution v2: Optimize
         */
        // idea
        // the problem says:
        // It does not matter what you leave beyond the returned k
        // thus, no need to switch the elements
        // just move retained element to front
        if (nums.length == 0) {
            return 0;
        }
        int ans = 0;
        for (int num : nums) {
            if (num != val) {
                nums[ans++] = num;
            }
        }
        return ans;
    }
}

File: LeetCode/Python/solutions/p0027.py
from typing import List


class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        #
        # Solution v1: Brute Force
        #
        # Runtime: 24 ms @ (beats) 98.60%
        # Memory Usage: 14.2 MB @ (beats) 48.87%
        #
        if len(nums) == 0:
            return 0
        elif len(nums) < 2:
            return 1 if nums[0] != val else 0

        left = 0
        right = len(nums) - 1
        while left <= right:
            if nums[left] == val:
                nums[left] = nums[right]
                right -= 1
            else:
                left += 1

        return left
